lib/Sema/TypeCheckPropertyWrapper.cpp

```diff
@@ -203,16 +203,57 @@ static ConstructorDecl *findInitialValueInit(
 static ConstructorDecl *findDefaultInit(ASTContext &ctx,
                                         NominalTypeDecl *nominal) {
   SmallVector<ConstructorDecl *, 2> defaultValueInitializers;
-  DeclName initName(ctx, DeclBaseName::createConstructor(),
-                    ArrayRef<Identifier>());
   SmallVector<ValueDecl *, 2> decls;
+  auto initName = DeclBaseName::createConstructor();
   nominal->lookupQualified(nominal, initName, NL_QualifiedDefault, decls);
   for (const auto &decl : decls) {
     auto init = dyn_cast<ConstructorDecl>(decl);
-    if (!init || init->getDeclContext() != nominal)
+    if (!init || init->getDeclContext() != nominal || init->isGeneric())
       continue;
 
-    defaultValueInitializers.push_back(init);
+    assert(init->hasParameterList());
+    // A constructor which does not have any parameters or where all the
+    // parameters have a default argument can be used to default initialize
+    // the property wrapper type.
+    //
+    // A constructor with no parameters will satisfy the check below.
+    bool allParamsHaveDefaultArg = llvm::all_of(
+        init->getParameters()->getArray(),
+        [](const ParamDecl *decl) { return decl->isDefaultArgument(); });
+
+    if (allParamsHaveDefaultArg) {
+      defaultValueInitializers.push_back(init);
+    }
+  }
+
+  if (defaultValueInitializers.size() > 1) {
+    // Let's find the single most specialized init. If we don't have one, then
+    // we'll diagnose later.
+    if (auto TC = static_cast<TypeChecker *>(ctx.getLazyResolver())) {
+      Optional<unsigned> bestIdx;
+      for (unsigned i = 1, n = defaultValueInitializers.size(); i != n; ++i) {
+        auto bestOrPrevIdx = bestIdx ? bestIdx.getValue() : i - 1;
+        auto firstInit =
+            cast<ValueDecl>(defaultValueInitializers[bestOrPrevIdx]);
+        auto secondInit = cast<ValueDecl>(defaultValueInitializers[i]);
+
+        switch (TC->compareDeclarations(nominal, firstInit, secondInit)) {
```

Comment:
I know I suggested this, but having gone down the path of https://github.com/apple/swift/pull/27322, I've convinced myself that's the better approach: rather than having another "pick the best declaration" implementation here, check whether any of the initializers fit the form... and then record the presence of such an initializer rather than which one was picked. Let ambiguity resolution and diagnostics be the responsibility of the expression type checker. I bet you could even generalize `findInitialValueInit` to cover the default-init case, and end up with a net deletion of code in the compiler. What do you think?

Reply:
I had the same thought yesterday... it sounds like a good idea! I think `findInitialValueInit` can already handle this because it requires all parameters to be either (1) default or (2) have the given argument label and we just care about (1).